### ISSUE TITLE
Error: Bad element kind BINDING_VARIABLE when using Java 15 'instanceof' with pattern matching.

### ISSUE THREAD

**Issue by mjoaquim12** [google/error-prone#2230]
### Description of the problem / feature request:

> Error during compilation:

```
[ERROR]      Please report this at https://github.com/google/error-prone/issues/new and include the following:
[ERROR]   
[ERROR]      error-prone version: 2.5.1
[ERROR]      BugPattern: NullAway
[ERROR]      Stack Trace:
[ERROR]      com.google.common.util.concurrent.UncheckedExecutionException: org.checkerframework.javacutil.BugInCF: bad element kind BINDING_VARIABLE
[ERROR]         at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
[ERROR]         at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
[ERROR]         at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
[ERROR]         at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4935)
[ERROR]         at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4941)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow.dataflow(DataFlow.java:146)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow.resultFor(DataFlow.java:276)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow.resultForExpr(DataFlow.java:253)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow.expressionDataflow(DataFlow.java:196)
[ERROR]         at com.uber.nullaway.dataflow.AccessPathNullnessAnalysis.getNullness(AccessPathNullnessAnalysis.java:119)
[ERROR]         at com.uber.nullaway.NullAway.nullnessFromDataflow(NullAway.java:1961)
[ERROR]         at com.uber.nullaway.NullAway.mayBeNullExpr(NullAway.java:1929)
[ERROR]         at com.uber.nullaway.NullAway.matchDereference(NullAway.java:2011)
[ERROR]         at com.uber.nullaway.NullAway.matchMemberSelect(NullAway.java:436)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.processMatchers(ErrorProneScanner.java:450)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMemberSelect(ErrorProneScanner.java:727)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMemberSelect(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:2293)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitMethodInvocation(TreeScanner.java:527)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:752)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1761)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitBinary(TreeScanner.java:644)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:514)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitBinary(TreeScanner.java:644)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:514)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitBinary(TreeScanner.java:643)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:514)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitBinary(TreeScanner.java:644)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:514)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBinary(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitReturn(TreeScanner.java:488)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:818)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1665)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:250)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:521)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1059)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:208)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:741)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:925)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:189)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:549)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:832)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:74)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:48)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
[ERROR]         at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:561)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:151)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:603)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
[ERROR]         at com.google.errorprone.scanner.Scanner.scan(Scanner.java:58)
[ERROR]         at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:43)
[ERROR]         at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:152)
[ERROR]         at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
[ERROR]         at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1421)
[ERROR]         at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1368)
[ERROR]         at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:960)
[ERROR]         at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:317)
[ERROR]         at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:176)
[ERROR]         at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:59)
[ERROR]         at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:45)
[ERROR]   Caused by: org.checkerframework.javacutil.BugInCF: bad element kind BINDING_VARIABLE
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitIdentifier(CFGBuilder.java:4038)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitIdentifier(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIdent.accept(JCTree.java:2427)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.getReceiver(CFGBuilder.java:2829)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitMethodInvocation(CFGBuilder.java:2583)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitMethodInvocation(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1761)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:3256)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:3352)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:3338)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:3352)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBinary(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBinary.accept(JCTree.java:2114)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitReturn(CFGBuilder.java:4257)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitReturn(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1665)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:3376)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.visitBlock(CFGBuilder.java:1526)
[ERROR]         at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1059)
[ERROR]         at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder$CFGTranslationPhaseOne.process(CFGBuilder.java:1681)
[ERROR]         at org.checkerframework.dataflow.cfg.CFGBuilder.build(CFGBuilder.java:260)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:130)
[ERROR]         at com.uber.nullaway.dataflow.DataFlow$2.load(DataFlow.java:97)
[ERROR]         at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
[ERROR]         at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
[ERROR]         at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
[ERROR]         at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
[ERROR]         ... 110 more
[ERROR]   Caused by: java.lang.Throwable
[ERROR]         at org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
[ERROR]         ... 151 more
```

### Feature requests: what underlying problem are you trying to solve with this feature?

> Compile code using error-prone and NullAway with Maven and JDK15 (--enable-preview)

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

> When using instanceof with pattern matching. Test project attached.
[Test.zip](https://github.com/google/error-prone/files/6109007/Test.zip)
`public boolean equals(Object obj) { return this == obj || obj instanceof Employee other && age == other.age; }
`

### What version of Error Prone are you using?

> 2.5.1

###  Have you found anything relevant by searching the web?

> No luck.
>
> Places to look:
> - GitHub issues: https://github.com/google/error-prone/issues
> - email threads on https://groups.google.com/forum/#!forum/error-prone-discuss

**Comment by cushon** [comment]
The Checker Framework dataflow component doesn't support Java 15 yet, I reported the crash here: https://github.com/typetools/checker-framework/issues/4410

**Comment by xeals** [comment]
The upstream issue has been fixed and released as part of version [3.13.0](https://github.com/typetools/checker-framework/releases/tag/checker-framework-3.13.0).
